### ISSUE TITLE
[FEATURE] Allow upgrade when there's VMs in the cluster that's can't be Live Migrated

### ISSUE THREAD

**Issue by spranta-devops** [harvester/harvester#6145]
**Is your enhancement related to a problem? Please describe.**
Since Harvester 1.2.2 the upgrade can only be started when all VMs are stopped that can not be migrated to another 
host. I have 5 Harvester Hosts, each of them is running a MinIO VM with PCI passthrough disks. These VMs can not be 
migrated to another host, because of the pci passthrough disks. I can not upgrade Harvester anymore without a long 
downtime of my MinIO, but it would be OK for MinIO if a single node (VM) goes down for some time.  

**Describe the solution you'd like**
Enhance the upgrade process by automatically stop non-migratable VMs when upgrading a host, and start them again after 
the reboot of the host. You can still display a warning with the list of the VMs.

**Describe alternatives you've considered**
Allow starting the upgrade despite the warning. Maybe with a "I know what I do"-checkbox. So far I've been monitored 
the upgrade process and stopped and started these VMs manually.

**Comment by innobead** [comment]
cc @votdev this seems related to what you are working on? Node maintenance mode strategy?

**Comment by votdev** [comment]
Looks like the feature improvement of the maintenance mode i was working on. Check this:
- https://github.com/harvester/harvester/issues/5069
- The related HEP: https://github.com/harvester/harvester/pull/5984
- The PR: https://github.com/harvester/harvester/pull/5328
- Documentation: https://github.com/harvester/docs/pull/523

**Comment by votdev** [comment]
@innobead I think we can close this issue as already implemented.

**Comment by spranta-devops** [comment]
The improved maintenance mode looks good to me.

Maybe I missed that, but the upgrade dialog is still blocking an upgrade, right? The linked issues didn't mention anything about the upgrade process and I couldn't find any changes regarding the upgrade process. 

See /pkg/webhook/resources/upgrade/validator.go:
```
        nonLiveMigratableVMNames, err := virtualmachineinstance.GetAllNonLiveMigratableVMINames(allVMIs, allNodes)
	if err != nil {
		return err
	}

	if len(nonLiveMigratableVMNames) > 0 {
		return werror.NewInvalidError(
			fmt.Sprintf("There are non-live migratable VMs that need to be shut off before initiating the upgrade: %s", strings.Join(nonLiveMigratableVMNames, ", ")), "",
		)
	}
```

**Comment by FrankYang0529** [comment]
> but the upgrade dialog is still blocking an upgrade, right? The linked issues didn't mention anything about the upgrade process and I couldn't find any changes regarding the upgrade process.

Yes, you're right. We should also consider whether the VM has `harvesterhci.io/maintain-mode-strategy` label. If it has, ignore it in `GetAllNonLiveMigratableVMINames` function.

**Comment by bk201** [comment]
> @innobead I think we can close this issue as already implemented.

This might need some investigation and more work. The current controller implementation in the maintenance path watches node changes and maintenance annotations. In the upgrade path, VMs on a node are live-migrated by adding a taint to an upgrading node: VMs are being migrated before Rancher drains a node. The timing to power on non-migratable VMs is also different (after a maintenance vs node is upgraded and rebooted).

**Comment by forbesguthrie** [comment]
What's being asked for here is that the upgrade process will allow upgrading nodes one by one.  This enables the admin to shutdown VMs which are application redundant across the cluster, the node to be upgraded, the VMs powered back on, then the next host to be selected for upgrade.  We may need to consider disabling other cluster functions during the upgrade to avoid customers pausing upgrades for extended periods.
Alternative approaches could be:
- to enable users to specify which VMs should be shutdown during an upgrade.
- pre-upgrade check for VMs that are obviously locked to hardware (e.g. passthrough), and ask the user for each VM what to do (pause upgrade, eject device, shutdown VM, power off VM) and the post-upgrade action (power-on or leave off)

**Comment by hoo29** [comment]
Until this feature is implemented, is there anything "unsupported" we can do that will allows an upgrade without downtime?

We have several clustered services all on replica 1 storage which means we currently have to shut them all down, even when we have the `harvesterhci.io/maintain-mode-strategy` set.

**Comment by rajha-korithrien** [comment]
I would like to add that our deployments would also greatly benefit from this feature.

We host an application that does its own clustering and high availability across 3 VMs on 3 different bare metal nodes.
The VMs are pinned to their bare metal host due to PCI passthrough of GPU resources.

Shutting down all 3 VMs at the same time is not a good option for us (as the upgrade process currently demands).
Shutting down each VM individually when its bare metal node gets updated, and then starting the VM back up after the node update is completed is an excellent option.

Our VMs have `harvesterhci.io/maintain-mode-strategy` set to `ShutdownAndRestartAfterDisable`

**Comment by ihcsim** [comment]
I think this is resolved by https://github.com/harvester/harvester/pull/8526 in 1.6. If the `restoreVM` option of the `upgrade-config` setting is set to `true`, Harvester turns off all non-migratable VMs on a node while attempting to pre-drain the node. Once the node is successfully upgraded, Harvester will try to restart those VMs.

**Comment by hoo29** [comment]
I think there is a still piece missing around waiting for the "health" of applications running on VMs before progressing with the upgrade. 

We have a clustered postgres across multiple VMs using single replica storage, each VM will need to be shutdown during an upgrade. When a postgres VM is turned back on, we don't want another one to be shutdown until all postgres replicas have fully caught up again. If another VM is shutdown before the replicas have caught up, it risks data loss.

This is similar for Minio and other clustered services, after a VM is turned off and on again we don't want further VMs turned off until the hosted application is healthy, quorum is reached, everything is caught up etc.

**Comment by ihcsim** [comment]
I don't think it's efficient to block the entire upgrade process to wait on per-node applications readiness. What you described is a common orchestration problem among stateful services in K8s. Normally, this is handled with sidecar hooks, and K8s' pod disruption budget and eviction API. I think it deserves its own separate issue for further discussion.

**Comment by hoo29** [comment]
Sure, I have created https://github.com/harvester/harvester/issues/8764

**Comment by Vicente-Cheng** [comment]
duplicated with #8980
